camel-k-runtime-health/src/main/java/org/apache/camel/k/health/HealthEndpoint.java

```diff
@@ -0,0 +1,130 @@
+import io.netty.handler.codec.http.HttpServerExpectContinueHandler;
+import io.netty.handler.codec.http.HttpVersion;
+import io.netty.handler.logging.LogLevel;
+import io.netty.handler.logging.LoggingHandler;
+import io.netty.util.AsciiString;
+import org.apache.camel.CamelContext;
+import org.apache.camel.ServiceStatus;
+import org.apache.camel.support.ServiceSupport;
+
+public class HealthEndpoint extends ServiceSupport {
+    private static final AsciiString CONTENT_TYPE = AsciiString.cached("Content-Type");
+    private static final AsciiString CONTENT_LENGTH = AsciiString.cached("Content-Length");
+    private static final byte[] OK = {'O', 'K'};
+    private static final byte[] KO = {'K', 'O'};
+
+    private final CamelContext context;
+    private final String bindHost;
+    private final int bindPort;
+    private final String path;
+
+    private EventLoopGroup bossGroup;
+    private EventLoopGroup workerGroup;
+    private Channel channel;
+
+    public HealthEndpoint(CamelContext context, String bindHost, int bindPort, String path) {
+        this.context = context;
+        this.bindHost = bindHost;
+        this.bindPort = bindPort;
+        this.path = path;
+    }
+
+    @Override
+    protected void doStart() throws Exception {
+        bossGroup = new NioEventLoopGroup(1);
+        workerGroup = new NioEventLoopGroup();
+
+        ServerBootstrap b = new ServerBootstrap();
+        b.group(bossGroup, workerGroup)
+            .channel(NioServerSocketChannel.class)
+            .handler(new LoggingHandler(LogLevel.DEBUG))
+            .childHandler(new ChannelInitializer<SocketChannel>() {
+                @Override
+                protected void initChannel(SocketChannel ch) throws Exception {
+                    ch.pipeline()
+                        .addLast(new HttpServerCodec())
+                        .addLast(new HttpServerExpectContinueHandler())
+                        .addLast(new Handler());
+                }
+            });
+
+        channel = b.bind(bindHost, bindPort).channel();
+    }
+
+    @Override
+    protected void doStop() throws Exception {
+        if (bossGroup != null) {
+            bossGroup.shutdownGracefully();
+        }
+        if (workerGroup != null) {
+            workerGroup.shutdownGracefully();
+        }
+    }
+
+    private class Handler extends SimpleChannelInboundHandler<HttpRequest> {
+        @Override
+        public void channelReadComplete(ChannelHandlerContext ctx) {
+            ctx.flush();
+        }
+
+        @Override
+        protected void channelRead0(ChannelHandlerContext ctx, HttpRequest msg) throws Exception {
+            HttpResponseStatus status;
+            ByteBuf content;
+
+            if (!Objects.equals(path, msg.uri())) {
+                status = HttpResponseStatus.NOT_FOUND;
+                content = Unpooled.wrappedBuffer(KO);
+            } else if (context.getStatus() == ServiceStatus.Started) {
```

Comment:
Is this true when routes are down for master-component/routepolicy?

Reply:
yes, context is always run only the consumer/route is stopped